source/comp/markv_codec.cpp

```diff
@@ -671,263 +991,1484 @@ class MarkvDecoder : public MarkvCodecBase {
-  assert(header_.markv_length_in_bits > 0);
+      if (multi_mtf_.HasValue(kMtfTypeComposite, data_type))
+        multi_mtf_.Insert(kMtfTypePointerToComposite, inst_.result_id);
+    }
+  }
 
-  if (header_.magic_number != kMarkvMagicNumber)
-    return vstate_.diag(SPV_ERROR_INVALID_BINARY)
-        << "MARK-V binary has incorrect magic number";
+  if (spvOpcodeGeneratesType(opcode)) {
+    if (opcode != SpvOpTypeFunction) {
+      multi_mtf_.Insert(kMtfTypeNonFunction, inst_.result_id);
+    }
+  }
 
-  // TODO(atgoo@github.com): Print version strings.
-  if (header_.markv_version != GetMarkvVersion())
-    return vstate_.diag(SPV_ERROR_INVALID_BINARY)
-        << "MARK-V binary and the codec have different versions";
+  const uint32_t descriptor = id_descriptors_.ProcessInstruction(inst_);
 
-  spirv_.reserve(header_.markv_length_in_bits / 2); // Heuristic.
-  spirv_.resize(5, 0);
-  spirv_[0] = kSpirvMagicNumber;
-  spirv_[1] = header_.spirv_version;
-  spirv_[2] = header_.spirv_generator;
+  multi_mtf_.Insert(GetMtfIdDescriptor(descriptor), inst_.result_id);
+}
 
-  while (reader_.GetNumReadBits() < header_.markv_length_in_bits) {
-    spv_parsed_instruction_t inst = {};
-    const spv_result_t decode_result = DecodeInstruction(&inst);
+uint64_t MarkvCodecBase::GetRuleBasedMtf() {
+  // This function is only called for id operands (but not result ids).
+  // TODO(atgoo@github.com) Clarify why SPV_OPERAND_TYPE_OPTIONAL_ID is not
+  // considered id type by spvIsIdType.
+  assert(spvIsIdType(operand_.type) ||
+         operand_.type == SPV_OPERAND_TYPE_OPTIONAL_ID);
+  assert(operand_.type != SPV_OPERAND_TYPE_RESULT_ID);
+
+  const SpvOp opcode = static_cast<SpvOp>(inst_.opcode);
+
+  // All operand slots which expect label id.
+  if ((inst_.opcode == SpvOpLoopMerge && operand_index_ <= 1) ||
+      (inst_.opcode == SpvOpSelectionMerge && operand_index_ == 0) ||
+      (inst_.opcode == SpvOpBranch && operand_index_ == 0) ||
+      (inst_.opcode == SpvOpBranchConditional &&
+       (operand_index_ == 1 || operand_index_ == 2 )) ||
+      (inst_.opcode == SpvOpPhi && operand_index_ >= 3 &&
+       operand_index_ % 2 == 1) ||
+      (inst_.opcode == SpvOpSwitch && operand_index_ > 0)) {
+    return kMtfLabel;
+  }
+
+  switch (opcode) {
+    case SpvOpFAdd:
+    case SpvOpFSub:
+    case SpvOpFMul:
+    case SpvOpFDiv:
+    case SpvOpFRem:
+    case SpvOpFMod:
+    case SpvOpFNegate: {
+      if (operand_index_ == 0)
+        return kMtfTypeFloatScalarOrVector;
+
+      return GetMtfIdOfType(inst_.type_id);
+    }
+
+    case SpvOpISub:
+    case SpvOpIAdd:
+    case SpvOpIMul:
+    case SpvOpSDiv:
+    case SpvOpUDiv:
+    case SpvOpSMod:
+    case SpvOpUMod:
+    case SpvOpSRem:
+    case SpvOpSNegate: {
+      if (operand_index_ == 0)
+        return kMtfTypeIntScalarOrVector;
+
+      return kMtfIntScalarOrVector;
+    }
+
+    // TODO(atgoo@github.com) Add OpConvertFToU and other opcodes.
```

Comment:
Ok.  When you add these opcodes, I guess you'll have to bump the Markv version though.  Ok with me.